### PR TITLE
DAOS-6431 dfuse: Add open_unlinked test, and allow for easier test methods.

### PR DESCRIPTION
**Pull request by ashleypittman**
…thods.

Create a in NLT adding tests and add three tests as methods.  This
allows isolation between tests, and finer grained control over what
is run and how it's reported.  This is how Python Unit Tests work.

Add command line options for running individual or all test methods,
and invoke all tests as part of regular testing.

Add three tests for open/unlinked files:
Where a open file has been unlinked by a rename operation
Where a open file has been renamed
Where a open file has been unlinked

Correct bug for first test where the wrong file was used in the
fstat reply, when performing dfs_ostat() check that the oid
is as expected by comparing it to the oid when the file was opened.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -3264,7 +3267,7 @@ dfs_osetattr(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
 	}
 
 out_stat:
-	rc = entry_stat(dfs, th, oh, obj->name, strlen(obj->name), stbuf);
+	rc = entry_stat(dfs, th, oh, obj->name, strlen(obj->name), obj, stbuf);
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c:3270]: im not sure i understand what you are trying to check for here.
If the object entry is wrong (someone changed it), you already updated the wrong entry in the update; and you are just doing the check after the fact?

Comment (ashleypittman) [reply src/client/dfs/dfs.c:3270]: I was going to write another test for that, and fix this at the same time.  I can roll it onto this PR though.

Comment (ashleypittman) [reply src/client/dfs/dfs.c:3270]: I think the fix for this is going to be a bit more involved.

File: utils/node_local_test.py [hunk 2]
```diff
@@ -934,6 +946,122 @@ def make_pool(daos):
...
+        """Test that fstat works on unlinked file"""
+        fname = os.path.join(self.dfuse.dir, 'unlinked')
+        ofd = open(fname, 'w')
+        print(os.fstat(ofd.fileno()))
+        os.unlink(fname)
+        try:
+            print(os.fstat(ofd.fileno()))
+            self.fail()
+        except FileNotFoundError:
+            print('Failed to fstat() renamed file')
```
Comment (kccain) [inline utils/node_local_test.py:1032]: suggest change the string to "Failed to fstat() unlinked file" ("renamed"->"unlinked").

Comment (ashleypittman) [reply utils/node_local_test.py]: Done, thanks.

File: src/client/dfs/dfs.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfs/dfs.c]: Done.  I added a check that no attempt is made to change the type of a file as well, although dfs_osetattr() is still not using transactions.

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -3314,14 +3319,26 @@ dfs_osetattr(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
 		return EPERM;
 	if ((obj->flags & O_ACCMODE) == O_RDONLY)
 		return EPERM;
+	if (flags & DFS_SET_ATTR_MODE)
+		if ((stbuf->st_mode & S_IFMT) != (obj->mode & S_IFMT))
+			return EINVAL;
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:3324]: This partially resolves [DAOS-5847](https://jira.hpdd.intel.com/browse/DAOS-5847), right?

**Review by daltonbohning — APPROVED** [review]

File: src/client/dfs/dfs.c (hunk 3 above)
Comment (ashleypittman) [reply src/client/dfs/dfs.c:3324]: Yes.  I was not aware of that ticket.

**Pull request merged**